File: batch-app/src/CardDemo.Batch.Core/Services/TransactionPostingService.cs
using CardDemo.Batch.Core.Models;

namespace CardDemo.Batch.Core.Services;

/// <summary>
/// Modernized batch logic from CBTRN02C transaction posting.
/// </summary>
public sealed class TransactionPostingService
{
    public TransactionPostingResult PostDailyTransactions(
        IEnumerable<TransactionRecord> dailyTransactions,
        IReadOnlyDictionary<string, CardCrossReference> cardCrossReferences,
        IDictionary<long, AccountRecord> accounts)
    {
        var posted = new List<TransactionRecord>();
        var rejected = new List<TransactionRecord>();

        foreach (var transaction in dailyTransactions)
        {
            if (!cardCrossReferences.TryGetValue(transaction.CardNumber, out var crossReference))
            {
                rejected.Add(transaction);
                continue;
            }

            if (!accounts.TryGetValue(crossReference.AccountId, out var account))
            {
                rejected.Add(transaction);
                continue;
            }

            if (!account.IsActive)
            {
                rejected.Add(transaction);
                continue;
            }

            account.CurrentBalance += transaction.Amount;
            posted.Add(new TransactionRecord
            {
                TransactionId = transaction.TransactionId,
                TypeCode = transaction.TypeCode,
                CategoryCode = transaction.CategoryCode,
                Description = transaction.Description,
                Amount = transaction.Amount,
                CardNumber = transaction.CardNumber,
                AccountId = crossReference.AccountId,
            });
        }

        return new TransactionPostingResult(posted, rejected);
    }
}

public sealed record TransactionPostingResult(
    IReadOnlyList<TransactionRecord> PostedTransactions,
    IReadOnlyList<TransactionRecord> RejectedTransactions);

File: batch-app-java/carddemo-batch-core/src/main/java/com/carddemo/batch/core/services/TransactionPostingService.java
package com.carddemo.batch.core.services;

import com.carddemo.batch.core.models.AccountRecord;
import com.carddemo.batch.core.models.CardCrossReference;
import com.carddemo.batch.core.models.TransactionPostingResult;
import com.carddemo.batch.core.models.TransactionRecord;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Modernized batch logic from CBTRN02C transaction posting.
 */
@Service
public class TransactionPostingService {

    public TransactionPostingResult postDailyTransactions(
            Iterable<TransactionRecord> dailyTransactions,
            Map<String, CardCrossReference> cardCrossReferences,
            Map<Long, AccountRecord> accounts) {
        List<TransactionRecord> posted = new ArrayList<>();
        List<TransactionRecord> rejected = new ArrayList<>();

        for (TransactionRecord transaction : dailyTransactions) {
            CardCrossReference crossReference = cardCrossReferences.get(transaction.getCardNumber());
            if (crossReference == null) {
                rejected.add(transaction);
                continue;
            }

            AccountRecord account = accounts.get(crossReference.getAccountId());
            if (account == null) {
                rejected.add(transaction);
                continue;
            }

            if (!account.isActive()) {
                rejected.add(transaction);
                continue;
            }

            account.setCurrentBalance(account.getCurrentBalance().add(transaction.getAmount()));
            posted.add(transaction.withAccountId(crossReference.getAccountId()));
        }

        return new TransactionPostingResult(List.copyOf(posted), List.copyOf(rejected));
    }
}
